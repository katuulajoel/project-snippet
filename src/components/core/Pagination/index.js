/* -------------------------------------------------------------------------- */
/*                            External dependencies                           */
/* -------------------------------------------------------------------------- */
import React, { useState, useEffect } from "react";
import { Pagination, PaginationItem, PaginationLink } from "reactstrap";
import styled from "styled-components";
import PropTypes from "prop-types";

/* -------------------------- Internal dependencies ------------------------- */
import Icon from "../Icon";

/* --------------------------- Component proptypes -------------------------- */
const proptypes = {
  pages: PropTypes.number,
  page: PropTypes.number,
  PageButtonComponent: PropTypes.any,
  onPageChange: PropTypes.func,
  previousText: PropTypes.string,
  nextPage: PropTypes.func,
  previousPage: PropTypes.func,
  canPreviousPage: PropTypes.bool,
  canNextPage: PropTypes.bool,
  loadMore: PropTypes.func,
};

const PaginationWrapper = ({
  pages,
  page,
  onPageChange,
  previousPage,
  nextPage,
  canPreviousPage,
  canNextPage,
  loadMore,
}) => {
  const [visiblePages, setVisiblePages] = useState([]);

  const getVisiblePages = (page, total) => {
    if (total < 7) {
      return filterPages([1, 2, 3, 4, 5, 6], total);
    } else {
      if (page % 5 >= 0 && page > 4 && page + 2 < total) {
        return [1, page - 1, page, page + 1, total];
      } else if (page % 5 >= 0 && page > 4 && page + 2 >= total) {
        return [1, total - 3, total - 2, total - 1, total];
      } else {
        return [1, 2, 3, 4, 5, total];
      }
    }
  };

  useEffect(() => {
    setVisiblePages(getVisiblePages(null, pages));
    changePage(page + 1);
  }, [pages]);

  useEffect(() => {
    const visiblePages = getVisiblePages(page, pages);
    setVisiblePages(filterPages(visiblePages, pages));
    loadMore(page);
  }, [page]);

  const filterPages = (visiblePages, totalPages) => {
    return visiblePages.filter((page) => page <= totalPages);
  };

  const changePage = (selectedPage) => {
    const activePage = page + 1;

    if (selectedPage === activePage) {
      return;
    }
    /* const visiblePages = getVisiblePages(selectedPage, pages);
        setVisiblePages(filterPages(visiblePages, pages)); */
    onPageChange(selectedPage - 1);
  };

  const activePage = page + 1;

  return (
    <Pagination aria-label="Page navigation example">
      <StyledPaginationItem
        onClick={() => {
          previousPage();
        }}
        disabled={canPreviousPage}
      >
        <PaginationLink href="#">
          <Icon name="arrow-left" size="sm" />
        </PaginationLink>
      </StyledPaginationItem>

      <div className="Table__visiblePagesWrapper" style={{ display: "flex" }}>
        {visiblePages.map((page, index, array) => {
          return array[index - 1] + 2 < page ? (
            <>
              <StyledPaginationItem
                key={page}
                active={activePage === page}
                onClick={() => changePage(page)}
              >
                <PaginationLink href="#">...{page}</PaginationLink>
              </StyledPaginationItem>
            </>
          ) : (
            <StyledPaginationItem
              key={page}
              active={activePage === page}
              onClick={() => changePage(page)}
            >
              <PaginationLink href="#">{page}</PaginationLink>
            </StyledPaginationItem>
          );
        })}
      </div>
      <StyledPaginationItem
        onClick={() => {
          nextPage();
        }}
        disabled={canNextPage}
      >
        <PaginationLink href="#">
          <Icon name="arrow-right" size="sm" />
        </PaginationLink>
      </StyledPaginationItem>
    </Pagination>
  );
};

const StyledPaginationItem = styled(PaginationItem)`
  &.active a.page-link {
    background-color: #062e64;
    border-color: #062e64;
  }
`;

PaginationWrapper.propTypes = proptypes;

export default PaginationWrapper;
