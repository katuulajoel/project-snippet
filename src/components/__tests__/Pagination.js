import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";
import Pagination from "../Pagination";

describe("Auth layout test", () => {
  it("Snapshot test for Login component", () => {
    const tree = renderer
      .create(
        <Router>
          <Pagination />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
