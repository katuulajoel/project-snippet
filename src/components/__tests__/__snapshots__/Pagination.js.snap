// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth layout test Snapshot test for Login component 1`] = `
<nav
  aria-label="Page navigation example"
  className=""
>
  <ul
    className="pagination"
  >
    <li
      className="sc-bdnxRM fERAOk page-item"
      onClick={[Function]}
    >
      <a
        className="page-link"
        href="#"
      >
        <i
          className="tg-ic-arrow-left tunga-ic-sz-sm "
        />
      </a>
    </li>
    <div
      className="Table__visiblePagesWrapper"
      style={
        Object {
          "display": "flex",
        }
      }
    />
    <li
      className="sc-bdnxRM fERAOk page-item"
      onClick={[Function]}
    >
      <a
        className="page-link"
        href="#"
      >
        <i
          className="tg-ic-arrow-right tunga-ic-sz-sm "
        />
      </a>
    </li>
  </ul>
</nav>
`;
